pkg/ccl/importccl/read_import_mysql.go

```diff
@@ -143,5 +150,197 @@ func mysqlToCockroach(
+	evalCtx *tree.EvalContext, id sqlbase.ID, name string, in *mysql.TableSpec,
+) (*sqlbase.TableDescriptor, error) {
+	if in == nil {
+		return nil, errors.Errorf("could not read definition for table %q (possible unsupoprted type?)", name)
+	}
+
+	time := hlc.Timestamp{WallTime: evalCtx.GetStmtTimestamp().UnixNano()}
+	priv := sqlbase.NewDefaultPrivilegeDescriptor()
+	desc := sql.InitTableDescriptor(id, defaultCSVParentID, name, time, priv)
+
+	colNames := make(map[string]string)
+	for _, raw := range in.Columns {
+		def, err := mysqlColToCockroach(raw.Name.String(), raw.Type)
+		if err != nil {
+			return nil, err
+		}
+		col, _, _, err := sqlbase.MakeColumnDefDescs(def, &tree.SemaContext{}, evalCtx)
+		if err != nil {
+			return nil, err
+		}
+		desc.AddColumn(*col)
+		colNames[raw.Name.String()] = col.Name
+	}
+
+	for _, raw := range in.Indexes {
+		var elems tree.IndexElemList
+		for _, col := range raw.Columns {
+			elems = append(elems, tree.IndexElem{Column: tree.Name(colNames[col.Column.String()])})
+		}
+		idx := sqlbase.IndexDescriptor{Name: raw.Info.Name.String(), Unique: raw.Info.Unique}
+		if err := idx.FillColumns(elems); err != nil {
+			return nil, err
+		}
+		if err := desc.AddIndex(idx, raw.Info.Primary); err != nil {
+			return nil, err
+		}
+	}
+
+	// AllocateIDs mutates its receiver. `return desc, desc.AllocateIDs()`
+	// happens to work in gc, but does not work in gccgo.
+	//
+	// See https://github.com/golang/go/issues/23188.
+	err := desc.AllocateIDs()
+	return &desc, err
+}
+
+func mysqlColToCockroach(name string, col mysql.ColumnType) (*tree.ColumnTableDef, error) {
+	def := &tree.ColumnTableDef{Name: tree.Name(name)}
+
+	var length, scale int
+
+	if col.Length != nil {
+		num, err := strconv.Atoi(string(col.Length.Val))
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse length for column %q", name)
+		}
+		length = num
+	}
+
+	if col.Scale != nil {
+		num, err := strconv.Atoi(string(col.Scale.Val))
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse scale for column %q", name)
+		}
+		scale = num
+	}
+
+	switch col.SQLType() {
+	case mysqltypes.Char:
+		def.Type = strOpt(coltypes.Char, length)
+	case mysqltypes.VarChar:
+		def.Type = strOpt(coltypes.VarChar, length)
+	case mysqltypes.Text:
+		def.Type = strOpt(coltypes.Text, length)
+	case mysqltypes.Blob:
+		def.Type = coltypes.Bytes
+	case mysqltypes.VarBinary:
+		def.Type = coltypes.Bytes
+	case mysqltypes.Binary:
+		def.Type = coltypes.Bytes
+
+	case mysqltypes.Int8:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Uint8:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Int16:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Uint16:
+		def.Type = coltypes.SmallInt
+	case mysqltypes.Int24:
+		def.Type = coltypes.Int
+	case mysqltypes.Uint24:
+		def.Type = coltypes.Int
+	case mysqltypes.Int32:
+		def.Type = coltypes.Int
+	case mysqltypes.Uint32:
+		def.Type = coltypes.Int
+	case mysqltypes.Int64:
+		def.Type = coltypes.BigInt
```

Comment:
Why are we using bigint here and not int4 or int2 above? Is it ok to just use int everywhere? This is hard.

Reply:
My worry about attempting to preserve these as close as possible to MySQL is that they may imply that the behavior will be the same as it was. Whereas if we change all the types to be the most canonical form then I think that would encourage old MySQL users to read the documentation on our types. My opinion is that the type aliases are primarily for Postgres ORMs. I'm happy to hear another argument, but if it were me going from mysql to cockroach, I would be slightly miffed that although the displayed type names were the same, their behavior was different. I would expect the type name to change if the behavior from mysql was changing.